### DISCUSSION TITLE
Can't fix it Unable to load DLL 'tdjson.dll' or one of its dependencies

### DISCUSSION THREAD

**Discussion by Primus-max** [egramtel/tdsharp#164] [Q&A] [answered]
In the development environment, everything was fine, the project worked. It's time to transfer to the server, got the following error:
unhandled exception. System.DllNotFoundException: Unable to load DLL 'tdjson.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)
    at TdLib.Bindings.WindowsBindings.td_json_client_create()
    at TdLib.Bindings.WindowsBindings.ClientCreate()
    at TdLib.TdJsonClient..ctor(ITdLibBindings bindings)
    at TdLib.TdClient..ctor(ITdLibBindings bindings)
    at TdLib.TdClient..ctor()
    at GetPostsFromTG.TGClient..ctor() in Z:\Programming\ProjectC#\GetPostsFromTG\TGClient.cs:line 27
    at Program.<Main>$(String[] args) in Z:\Programming\ProjectC#\GetPostsFromTG\Program.cs:line 24
    at Program.<Main>(String[] args)
I'm new to programming and can't figure this out on my own. I would really appreciate help.

**Comment by ForNeVeR** [comment]
What exactly have you transferred to server? Is `tdlibjson.dll` among those files?

**Reply by Primus-max** [reply]
Thanks, I've figured it out. Yeah, it was about assembly. More precisely, no matter what assembly was made, there was still an error, then on the server for which the program installed the library through the source project file in visual studio. Although it should work differently, and everything will be key in the assembly.

**Reply by ForNeVeR** [reply]
Your comment is confusing. Normally, all the dependencies should be packed by `dotnet pack`. Probably you're trying to deploy a result of `dotnet build`?

**Comment by Primus-max** [comment]
I was also confused, when building I was told to build with all the dependencies, and the necessary library was in the build, but there was still an error. Next, I just ran the exe file. Of course, the mistake is in my actions, and I would like to understand in what actions.

**Reply by ForNeVeR** [reply]
So, are you able to run the resulting `.exe` file or not?

**Reply by Primus-max** [reply]
Yes, but I had to deploy a development environment on a server where I was going to use the program and install through the Nuget library, build and build the project. In fact, the problem remains.

**Reply by ForNeVeR** [reply]
Could you please explain _the problem_ in more detail?
1. What exact set of binaries do you deploy?
2. What's the result of this deployment?
3. What steps did you take as a temporary solution?

**Reply by Primus-max** [reply]
I apologize, I read it and forgot to answer. 
I tried to assemble differently, tried all the combinations of assemblies. The last build option was: Deployment mode: Framework-dependent, Target runtime: win-64. Target Framwork: .NET 6.0. The error said Unable to load DLL 'tdjson.dll'. For the solution, I deployed Visual Studio and ran it as a project, installed the TdLib and TdApi libraries. The development machine and the machine for use are different.

**Reply by ForNeVeR** [reply]
> For the solution, I deployed Visual Studio

I am sorry, but here you lost me. What does that mean, "deployed Visual Studio and ran it as a project"?

**Comment by Primus-max** [comment]
Opened the project on the server through visual studio, and installed through the Nuget library. And then dotnet build and then from the bin folder I was able to run myProgram.exe

**Comment by ForNeVeR** [comment, marked as answer]
Normally, if you deploy `tdjson.dll` alongside your program, it all should work correctly.

Note, though, that Windows build of `tdjson.dll` deployed as a part of tdlib.native (which you may or may not use, this is your choice) may depend on Visual Studio C++ runtime, that may have to be installed separately (that's [mentioned in the documentation](https://github.com/ForNeVeR/tdlib.native#library-dependencies)).

For investigation of such cases, I recommend using [Process Monitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon): set it up to filter file activity of your process and look for failed file opening attempts. That will help to determine which DLLs it looks for, and at what paths. If there's a failed query for `vcruntime*.dll`, that means you have no required runtime version installed. If it fails to load something like `api-ms-*.dll`, then that means UCRT is not installed. The latter usually comes with Windows updates, but god that's messy.

**Comment by Primus-max** [comment]
I understand roughly what you are talking about, I think I will still face
this problem since I will have to work with your library. Thanks for the
time.

вс, 2 апр. 2023 г. в 18:48, Friedrich von Never ***@***.***>:
